src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPopAndLockDatasetStep.java

```diff
@@ -0,0 +1,97 @@
+package bio.terra.service.snapshot.flight.delete;
+
+import bio.terra.common.exception.RetryQueryException;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.dataset.exception.DatasetLockException;
+import bio.terra.service.dataset.exception.DatasetNotFoundException;
+import bio.terra.service.dataset.flight.DatasetWorkingMapKeys;
+import bio.terra.service.iam.AuthenticatedUserRequest;
+import bio.terra.service.snapshot.Snapshot;
+import bio.terra.service.snapshot.SnapshotService;
+import bio.terra.service.snapshot.exception.SnapshotNotFoundException;
+import bio.terra.service.snapshot.flight.SnapshotWorkingMapKeys;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import java.util.UUID;
+
+public class DeleteSnapshotPopAndLockDatasetStep implements Step {
+
+  private SnapshotService snapshotService;
+  private DatasetService datasetService;
+  private UUID snapshotId;
+  private AuthenticatedUserRequest authenticatedUserRequest;
+  private boolean sharedLock;
+
+  public DeleteSnapshotPopAndLockDatasetStep(
+      UUID snapshotId,
+      SnapshotService snapshotService,
+      DatasetService datasetService,
+      AuthenticatedUserRequest authenticatedUserRequest,
+      boolean sharedLock) {
+    this.snapshotId = snapshotId;
+    this.snapshotService = snapshotService;
+    this.datasetService = datasetService;
+    this.authenticatedUserRequest = authenticatedUserRequest;
+    this.sharedLock = sharedLock;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) {
+    FlightMap map = context.getWorkingMap();
+    Snapshot snapshot;
+
+    // Confirm that snapshot exists and populate info about snapshot to inform optional steps
+    try {
+      snapshot = snapshotService.retrieve(snapshotId);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_EXISTS, true);
+      boolean hasGoogleProject =
+          snapshot.getProjectResource() != null
+              && snapshot.getProjectResource().getGoogleProjectId() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_GOOGLE_PROJECT, hasGoogleProject);
+      boolean hasAzureStorageAccount = snapshot.getStorageAccountResource() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_AZURE_STORAGE_ACCOUNT, hasAzureStorageAccount);
```

Comment:
@samanehsan  - Are these checks for hasAzureStorageAccount and hasGoogleProject going to work with the changes in DR-2107? Since I don't necessarily know if the source dataset still exists, I'm trying to infer the platform based on whether these properties exist on the snapshot.

Reply:
Yes, `snapshot.getProjectResource()` would be null for Azure snapshots with the changes in DR-2107

Reply:
Hmm actually I think the GCP steps might fail since the snapshot wouldn't have a dataset in BigQuery to delete. Would it make sense to temporarily set `hasGoogleProject` to false if `hasAzureStorageAccount` is true? Or combine our PRs into a feature branch?

Reply:
Huh, that's a good point. Tests aren't failing, but maybe I should be suspicious of that…

Reply:
Ok - Results of investigation: 
(1) We're still creating the source dataset's views, so that's why were able to successfully delete ACLs
(2) The delete dataset call on the snapshot's BQ dataset was failing silently, so I've added a log.